File: scripts/regen_og_pngs.js
#!/usr/bin/env node
/**
 * Batch-convert all per-skill OG SVGs to PNG using sharp (librsvg bundled).
 * Works on Windows, macOS, Linux without any system-level Cairo install.
 *
 * Usage (from repo root):
 *   npm install --no-save sharp
 *   node scripts/regen_og_pngs.js
 */
const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const OG_DIR = 'docs/og';
const DOCS_DIR = 'docs';
const AOV_DIR = path.join(DOCS_DIR, 'assets', 'ascension-overdrive');
const SKIP = new Set(['social-preview.svg']);
const W = 1200, H = 630;

function rglob(dir) {
  const out = [];
  for (const f of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, f.name);
    if (f.isDirectory()) out.push(...rglob(full));
    else if (f.name.endsWith('.svg') && !SKIP.has(f.name)) out.push(full);
  }
  return out;
}

async function assetDataUri(assetPath) {
  // Re-encode WebP medallions as PNG before handing the SVG to librsvg.  This
  // avoids both local `/assets/...` resolution failures and Cairo/WebP loader
  // differences that otherwise produce a valid PNG with an empty medallion.
  const png = await sharp(assetPath).png().toBuffer();
  return `data:image/png;base64,${png.toString('base64')}`;
}

async function inlineRasterAssets(svgSource) {
  if (!fs.existsSync(AOV_DIR)) return svgSource;
  const files = fs.readdirSync(AOV_DIR)
    .filter((name) => /^aov4-.*-hero\.webp$/.test(name))
    .sort();
  let out = svgSource;
  for (const name of files) {
    const href = `/assets/ascension-overdrive/${name}`;
    if (!out.includes(href)) continue;
    const dataUri = await assetDataUri(path.join(AOV_DIR, name));
    out = out.split(href).join(dataUri);
  }
  return out;
}

(async () => {
  if (!fs.existsSync(OG_DIR)) {
    console.error('ERROR: docs/og/ not found. Run from repo root.');
    process.exit(1);
  }
  const svgs = rglob(OG_DIR);
  if (!svgs.length) { console.log('No SVGs found.'); return; }
  console.log(`Converting ${svgs.length} SVGs → PNGs (${W}×${H})...`);
  let ok = 0, errors = [];
  for (const svg of svgs) {
    const png = svg.replace(/\.svg$/, '.png');
    try {
      const source = await inlineRasterAssets(fs.readFileSync(svg, 'utf8'));
      await sharp(Buffer.from(source)).resize(W, H).png().toFile(png);
      console.log(`  PNG: ${png}`);
      ok++;
    } catch (e) {
      errors.push({ svg, msg: e.message });
      console.error(`  ERR: ${svg} — ${e.message}`);
    }
  }
  console.log(`\nGenerated ${ok}/${svgs.length} PNG(s).`);
  if (errors.length) process.exit(1);
})();

File: scripts/regen_og_pngs.py
#!/usr/bin/env python3
"""Batch-convert all per-skill OG SVGs to PNG using cairosvg (sequential).

Run from the repo root:
    pip install cairosvg
    python scripts/regen_og_pngs.py

Skips social-preview.svg (different 1280x640 dimensions).
"""
import base64
import sys
from io import BytesIO
from pathlib import Path

OG_W = 1200
OG_H = 630
SKIP = {"social-preview.svg"}
DOCS_DIR = Path("docs")


def _asset_data_uri(path: Path) -> str:
    """Return a raster-safe data URI for an SVG image asset.

    The committed OG SVGs use root-relative `/assets/...webp` links.  When a
    batch converter runs against local files those links are not guaranteed to
    resolve, and some Cairo builds cannot decode WebP even after inlining.  A
    PNG data URI is the most portable raster input; fall back to typed WebP if
    Pillow is unavailable.
    """
    try:
        from PIL import Image

        with Image.open(path) as img:
            buf = BytesIO()
            img.save(buf, format="PNG")
        return "data:image/png;base64," + base64.b64encode(buf.getvalue()).decode("ascii")
    except Exception:
        mime = "image/webp" if path.suffix.lower() == ".webp" else "application/octet-stream"
        return f"data:{mime};base64," + base64.b64encode(path.read_bytes()).decode("ascii")


def inline_raster_assets(svg_content: str, docs_dir: Path = DOCS_DIR) -> str:
    """Inline root-relative docs assets before rasterizing an OG SVG."""
    asset_dir = docs_dir / "assets" / "ascension-overdrive"
    for path in sorted(asset_dir.glob("aov4-*-hero.webp")):
        href = "/" + path.relative_to(docs_dir).as_posix()
        if href in svg_content:
            svg_content = svg_content.replace(href, _asset_data_uri(path))
    return svg_content


def main() -> None:
    try:
        import cairosvg
    except ImportError:
        print("ERROR: cairosvg not installed. Run: pip install cairosvg", file=sys.stderr)
        sys.exit(1)

    og_dir = Path("docs/og")
    if not og_dir.exists():
        print(f"ERROR: {og_dir} not found. Run from repo root.", file=sys.stderr)
        sys.exit(1)

    svgs = sorted(p for p in og_dir.rglob("*.svg") if p.name not in SKIP)
    if not svgs:
        print("No SVGs found in docs/og/")
        return

    png_count = 0
    errors = []
    for svg_path in svgs:
        png_path = svg_path.with_suffix(".png")
        try:
            raster_svg = inline_raster_assets(svg_path.read_text(encoding="utf-8"))
            cairosvg.svg2png(
                bytestring=raster_svg.encode("utf-8"),
                write_to=str(png_path),
                output_width=OG_W,
                output_height=OG_H,
            )
            print(f"  PNG: {png_path.relative_to(og_dir.parent)}")
            png_count += 1
        except Exception as e:
            errors.append((svg_path, e))
            print(f"  ERR: {svg_path} — {e}", file=sys.stderr)

    print(f"\nGenerated {png_count}/{len(svgs)} PNG(s).")
    if errors:
        print(f"{len(errors)} error(s) — see above.", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
